src/Microsoft.AspNetCore.Hosting/WebHostExtensions.cs

```diff
@@ -95,7 +153,16 @@ private static void Run(this IWebHost host, CancellationToken token, string shut
                 },
                 applicationLifetime);
 
-                applicationLifetime.ApplicationStopping.WaitHandle.WaitOne();
+                var waitForStop = new TaskCompletionSource<object>();
+                applicationLifetime.ApplicationStopping.Register(() =>
+                {
+                    // Prevent waitForStop from continuing on the Stopping thread to avoid deadlocks.
+                    Task.Run(() => { waitForStop.TrySetResult(null); });
```

Comment:
Undo and use the brand new TCS API.`new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);`